### DISCUSSION TITLE
Problem in multi-node training

### DISCUSSION THREAD

**Discussion by nickKyr** [Lightning-AI/pytorch-lightning#7275] [DDP / multi-GPU / multi-node] [answered]
Hello pytorch-lightning community,

my training hangs when training on multi-nodes; on single node with multiple GPUs runs fine :/
It baffles me that although the global rank ID seems right, the member output has 4 instead of 8 in the denominator.
Since I run in a slurm environment, do I have to add the SLURMEnvironment plugin in the Trainer? I tried to add it alongside the DDPPlugin but it was not accepted (Found invalid type for plugin <class 'pytorch_lightning.plugins.environments.slurm_environment.SLURMEnvironment'>. Expected a precision or training type plugin)

The job submission file has the corresponding lines:
#SBATCH --gres=gpu:4
#SBATCH --nodes=2
#SBATCH --exclusive

srun --ntasks=8 python3 coolModel.py 2>&1 | tee log.train

I attach the output and the code below...
pytorch version:  1.8.1
pytorch-lightning version:  1.2.4

Cheers,
Nikos

[multNodeTraining.txt](https://github.com/PyTorchLightning/pytorch-lightning/files/6398914/multNodeTraining.txt)

###########python code
import os
import torch
from torch.nn import functional as F
from torch.utils.data import DataLoader
from torchvision.datasets import MNIST
import torchvision.transforms as transforms

import pytorch_lightning as pl
from pytorch_lightning import Trainer
#from test_tube import Experiment

class database(pl.LightningDataModule):

    def __init__(self):
        super().__init__()

    #def setup(self, stage=None):
    def train_dataloader(self):
        return DataLoader(MNIST(os.getcwd(), train=True, download=True, transform=transforms.ToTensor()), batch_size=32)

    def val_dataloader(self):
        return DataLoader(MNIST(os.getcwd(), train=True, download=True, transform=transforms.ToTensor()), batch_size=32)

    def test_dataloader(self):
        return DataLoader(MNIST(os.getcwd(), train=True, download=True, transform=transforms.ToTensor()), batch_size=32)

class CoolModel(pl.LightningModule):

    def __init__(self):
        super(CoolModel, self).__init__()
        # not the best model...
        self.l1 = torch.nn.Linear(28 * 28, 10)

    def forward(self, x):
        return torch.relu(self.l1(x.view(x.size(0), -1)))

    def my_loss(self, y_hat, y):
        return F.cross_entropy(y_hat, y)

    def training_step(self, batch, batch_nb):
        x, y = batch
        y_hat = self.forward(x)
        return {'loss': self.my_loss(y_hat, y)}

    def validation_step(self, batch, batch_nb):
        x, y = batch
        y_hat = self.forward(x)
        return {'val_loss': self.my_loss(y_hat, y)}

    def validation_end(self, outputs):
        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
        return {'avg_val_loss': avg_loss}

    def configure_optimizers(self):
        return [torch.optim.Adam(self.parameters(), lr=0.02)]

if __name__=='__main__':

    dm = database()
    model = CoolModel()
    #exp = Experiment(save_dir=os.getcwd())
    #checkp1 = pl.callbacks.ModelCheckpoint(
    #    monitor='loss', save_top_k=2, dirpath='./', mode='min', save_last=True
    #    ) 
    trainer = Trainer(
            max_epochs=10,
            gpus=4, num_nodes=2, accelerator='ddp',
            plugins= pl.plugins.DDPPlugin(find_unused_parameters=False), #pl.plugins.SLURMEnvironment],
            #progress_bar_refresh_rate=0
            )
    # train on 32 gpus across 4 nodes (make sure to submit appropriate SLURM job)
    # trainer = Trainer(experiment=exp, max_nb_epochs=1, gpus=[0, 1, 2, 3, 4, 5, 6, 7], nb_gpu_nodes=4)
    trainer.fit(model, dm)
    # view tensorflow logs
    print(f'View tensorboard logs by running\ntensorboard --logdir {os.getcwd()}')
    print('and going to http://localhost:6006 on your browser')

**Comment by awaelchli** [comment, marked as answer]
Hello Nikos

Do you have 8 gpus in the node? I think it must match gres. 
Don't you also need to specify how many tasks per node in the SBATCH directive? [1]
Also, I notice some unsupported Trainer arguments in your script. It should be:
`trainer = Trainer(max_epochs=1, gpus=[0, 1, 2, 3, 4, 5, 6, 7], num_nodes=4)
`
Make sure this script actually runs on CPU first before going to the cluster 😅 

Totally no slurm expert here, just looking at your script with one eye closed. 


[1] https://pytorch-lightning.readthedocs.io/en/latest/clouds/cluster.html#slurm-managed-cluster

**Reply by nickKyr** [reply]
Hello Adrian,

thanks for your reply.

No, I have 4 GPUs in each node, so when requesting 2 nodes, I should be getting 8 GPUs in total, based on the link you sent me, which has the following: 
```
# train on 32 GPUs across 4 nodes
trainer = Trainer(gpus=8, num_nodes=4, accelerator='ddp')
```

I was quite certain that num_nodes is the GPUs per node and not the total number, otherwise, you get an error.

I tried to specify the tasks per node, but I got the following error:

> sbatch: error: Job rejected: Please do not specify cores/CPUs/tasks for GPU jobs. This is controlled in proportion to GPUs requested.
> sbatch: error: Batch job submission failed: Unspecified error

I then tried to specify the tasks per node in the srun command with a flag but the code stacks at the same stage:
`srun --ntasks=8 --tasks-per-node=4 python3 coolModel.py 2>&1 | tee log.train`

I thought `n_gpus=4` is equivalent with `n_gpus=[0, 1, 2, 3]`. Nevertheless, I tried the list format but the result is the same...

In principle, if the total number of GPUs or CPUs which is requested by the code doesn't match the one specified in the job submission file, you get an error, but this is not the case here.

Not sure what is missing...

Thanks,
Nikos

**Reply by awaelchli** [reply]
I can guarantee that the gpus argument applies per node. So it's important that you set either gpus=4 or (gpus=[0, 1, 2, 3]) but you had 8 which would certainly not work if your nodes only have 4 each. 

> I thought n_gpus=4 is equivalent with n_gpus=[0, 1, 2, 3]. Nevertheless, I tried the list format but the result is the same...

it's the same, it does not matter.

try `Trainer(gpus=4, num_nodes=2, accelerator="ddp", ...)`

and 

```
#SBATCH --gres=gpu:4
#SBATCH --nnodes=2
#SBATCH --ntasks-per-node=4 (maybe as you said it does not work)
```

and if that doesn't work try a dummy pytorch script to see that you can launch on multiple nodes successfully.

**Reply by nickKyr** [reply]
Hi,

if I use the same job submission file to run a simplistic python script on 2 nodes with 4 GPUs each, the message is printed 8 times as it should be:
```
if __name__=='__main__':
    print("Run on GPU" )
```
will give:
_Run on GPU
Run on GPU
Run on GPU
Run on GPU
Run on GPU
Run on GPU
Run on GPU
Run on GPU_
However, it seems like pytorch can only identify the gpus of the current node: `torch.cuda.device_count()` will output 4 and it is not so trivial to get the global GPU rank, as `torch.cuda.current_device()` is always outputting 0. [This](https://github.com/PyTorchLightning/pytorch-lightning/blob/6be0a859db93989768f8bc845ee60e9c73f1e809/pytorch_lightning/utilities/distributed.py#L55-L61) post worked for me:
```
import os
import torch

def _get_rank() -> int:
    rank_keys = ('RANK', 'SLURM_PROCID', 'LOCAL_RANK')
    for key in rank_keys:
        rank = os.environ.get(key)
        if rank is not None:
            return int(rank)
    return 0

if __name__=='__main__':
    print("Number of GPUs: ", torch.cuda.device_count() )
    print("Device: ", _get_rank() )
```
and the output I get is:
_Number of GPUs:  4
Device:  7
Number of GPUs:  4
Device:  3
Number of GPUs:  4
Device:  6
Number of GPUs:  4
Device:  4
Number of GPUs:  4
Device:  5
Number of GPUs:  4
Device:  0
Number of GPUs:  4
Device:  1
Number of GPUs:  4
Device:  2_
I feel like the root of the problem is the same as in my original post where the denominator in MEMBER was always 4 (see multNodeTraining.txt)

**Reply by awaelchli** [reply]
Hi
trainer.global_rank will give you the global process id (0 - 7 in your case). 
trainer.local_rank gives the local id within the node (0 - 3 in your case)
trainer.node_rank will give you the id of the node (0-1 in your case)
trainer.world_size gives you the total number of devices across all nodes (8 in your case)

**Reply by nickKyr** [reply]
Thanks for your reply.
Indeed, `world_size` is used in plugins/training_type/ddp.py:222:
`log.info(f"initializing ddp: GLOBAL_RANK: {global_rank}, MEMBER: {global_rank + 1}/{world_size}")`
However, in my case it outputs 4 instead of 8 and based on the above minimal pytorch output where all 8 GPUs seem to be utilised, I don't think it is a setting in the job submission that is missing...

**Reply by awaelchli** [reply]
oh and I forgot to say, 1.2.4 had known issues with slurm that we fixed. Use Lightning 1.2.10. Sorry, should have mentioned that. The world size is pulled from the environment. You still have to set num_nodes and gpus argument correctly in the Trainer even if you set it in the batch script for slurm.

**Reply by nickKyr** [reply]
Hi,
thanks for letting me know about that.

Indeed now it's better (v1.2.10) but not perfect (running on 2 GPU nodes, 4 GPUs each):
_GPU available: True, used: True
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
TPU available: False, using: 0 TPU cores
Multi-processing is handled by Slurm.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
initializing ddp: GLOBAL_RANK: 3, MEMBER: 4/8
initializing ddp: GLOBAL_RANK: 1, MEMBER: 2/8
initializing ddp: GLOBAL_RANK: 0, MEMBER: 1/8
initializing ddp: GLOBAL_RANK: 2, MEMBER: 3/8
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
initializing ddp: GLOBAL_RANK: 5, MEMBER: 6/8
initializing ddp: GLOBAL_RANK: 4, MEMBER: 5/8
initializing ddp: GLOBAL_RANK: 6, MEMBER: 7/8
initializing ddp: GLOBAL_RANK: 7, MEMBER: 8/8
Set SLURM handle signals.
Set SLURM handle signals.
Set SLURM handle signals.
Set SLURM handle signals.
Set SLURM handle signals.
Set SLURM handle signals.
Set SLURM handle signals.
Set SLURM handle signals._

Although the `GLOBAL_RANK` and `MEMBER` seem correct, the `CUDA_VISIBLE_DEVICES` are not and I get the following error for ranks 4 to 7:
  _File "/lib/python3.8/site-packages/pytorch_lightning/accelerators/accelerator.py", line 73, in start_training
    self.training_type_plugin.start_training(trainer)
  File "/lib/python3.8/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 114, in start_training
    self._results = trainer.run_train()
  File "/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 607, in run_train
    self.run_sanity_check(self.lightning_module)
  File "/lib/python3.8/site-packages/pytorch_lightning/trainer/trainer.py", line 854, in run_sanity_check
    self.reset_val_dataloader(ref_model)
  File "/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py", line 364, in reset_val_dataloader
    self.num_val_batches, self.val_dataloaders = self._reset_eval_dataloader(model, 'val')
  File "/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py", line 317, in _reset_eval_dataloader
    dataloaders = [self.auto_add_sampler(dl, shuffle=False) for dl in dataloaders if dl is not None]
  File "/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py", line 317, in <listcomp>
    dataloaders = [self.auto_add_sampler(dl, shuffle=False) for dl in dataloaders if dl is not None]
  File "/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py", line 113, in auto_add_sampler
    sampler = self._get_distributed_sampler(dataloader, shuffle)
  File "/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py", line 188, in _get_distributed_sampler
    sampler = DistributedSampler(dataloader.dataset, **kwargs)
  File "/lib/python3.8/site-packages/torch/utils/data/distributed.py", line 71, in __init__
    raise ValueError(
ValueError: Invalid rank 5, rank should be in the interval [0, 3]_

**Reply by awaelchli** [reply]
> LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]

if your node has 4 gpus then lightning will always print` CUDA_VISIBLE_DEVICES: [0,1,2,3]`, that's ok :) It's just user feedback how many devices are in the machine and visible to PL to select from. 

For the local rank messages, I just noticed this is a display bug. It should read:

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
LOCAL_RANK: 1 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
LOCAL_RANK: 2 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
LOCAL_RANK: 3 - CUDA_VISIBLE_DEVICES: [0,1,2,3]
...
I submitted a quick fix for that #7395

Regarding the error you get, no idea yet but I will try to do some guess work (I don't currently have a slurm cluster, but maybe soon).
 
If it's not too much work for you, could you print these environment variables right before trainer.fit and show me the output?

```python
print(os.environ["SLURM_PROCID"])
print(os.environ["SLURM_LOCALID"])
print(os.environ["SLURM_NODEID"])
```
thanks a lot

**Reply by nickKyr** [reply]
thanks for the reply. The output is shown below, but it is only for the first GPU node, I got the same error afterwards...

the PROCID is:  4
the LOCALID is:  0
the NODEID is:  1
the PROCID is:  6
the LOCALID is:  2
the NODEID is:  1
the PROCID is:  5
the LOCALID is:  1
the NODEID is:  1
the PROCID is:  7
the LOCALID is:  3
the NODEID is:  1

**Reply by awaelchli** [reply]
> but it is only for the first GPU node

did you mean to write second node? because slurm is telling us here this is node id 1 which means the second node. 

interesting, you are only able to print this for the second node (node id 1) and then the error occurs. 
any idea what happens on node 0?

**Reply by nickKyr** [reply]
Yes, I meant the second node, sorry... It's interesting that although the PROCIDs printed are greater than 3, the errors are about them:
ValueError: Invalid rank 4, rank should be in the interval [0, 3]
ValueError: Invalid rank 5, rank should be in the interval [0, 3]
ValueError: Invalid rank 6, rank should be in the interval [0, 3]
ValueError: Invalid rank 7, rank should be in the interval [0, 3]
No idea what's going on, that's why I posted it :) Even a single print statement on the top of the __main__ function (the main function originally posted 1 week ago), is printed 4 times instead of 8. So it seems like only 1 GPU node is creating processes...

**Reply by awaelchli** [reply]
aha! maybe the world size is incorrectly set to 4 instead of 8 and then the distributed sampler complains that the ranks should be in range [0, 4]. 

I have an idea. Maybe it's related to https://github.com/PyTorchLightning/pytorch-lightning/issues/7007
and because you pass in the plugin manually you also have to set num_nodes manually. Let's try this:

```
Trainer(
        max_epochs=10,
        gpus=4, num_nodes=2, accelerator='ddp',
        plugins= pl.plugins.DDPPlugin(
             num_nodes=2, # ADDED
             find_unused_parameters=False
        )
)

print(trainer.global_rank, trainer.world_size, os.environ["SLURM_NTASKS"], trainer.training_type_plugin.num_processes, trainer.training_type_plugin.num_nodes, trainer.training_type_plugin.num_processes)
```

should print
`X 8 8 2 4`

**Reply by nickKyr** [reply]
it works now! Specifying the number of nodes in the DDPPlugin was needed:

2 8 8 4 2 4
the PROCID is:  2
the LOCALID is:  2
the NODEID is:  0

5 8 8 4 2 4
the PROCID is:  5
the LOCALID is:  1
the NODEID is:  1

1 8 8 4 2 4
the PROCID is:  1
the LOCALID is:  1
the NODEID is:  0

6 8 8 4 2 4
the PROCID is:  6
the LOCALID is:  2
the NODEID is:  1

4 8 8 4 2 4
the PROCID is:  4
the LOCALID is:  0
the NODEID is:  1

7 8 8 4 2 4
the PROCID is:  7
the LOCALID is:  3
the NODEID is:  1

3 8 8 4 2 4
the PROCID is:  3
the LOCALID is:  3
the NODEID is:  0

0 8 8 4 2 4
the PROCID is:  0
the LOCALID is:  0
the NODEID is:  0

I even tested this on my actual application which is a much larger network and it works!
I had seen some similar posts, but not [7007](https://github.com/PyTorchLightning/pytorch-lightning/issues/7007)
Thank you Adrian!

**Reply by awaelchli** [reply]
AMAZING!
Ok, great. We know that adding the num_nodes in two places is not ideal. The linked issue will make sure we can infer it directly from the trainer, then there should be no inconsistency!

**Comment by tsikup** [comment]
I want to revisit this discussion because I find my self in a similar situation that I can't get out of.
I submit a SLURM job with 2 nodes, 4 gpus per node and 4 tasks per node (1 task per gpu) as suggested by the lightning docs [here](https://pytorch-lightning.readthedocs.io/en/latest/clouds/cluster_advanced.html). 

```
#SBATCH --nodes=2
#SBATCH --ntasks-per-node=4
#SBATCH --gpus-per-node=A100:4
```

with `devices=4` and `num_nodes=2` in the Trainer.

I get the following error although the first 4 ranks (gpus) have been initialized as follows
`Initializing distributed: GLOBAL_RANK: 2, MEMBER: 3/8`
`Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/8`
`Initializing distributed: GLOBAL_RANK: 1, MEMBER: 2/8`
`Initializing distributed: GLOBAL_RANK: 3, MEMBER: 4/8`

Can you help me?

```
Traceback (most recent call last):
  File "/local/tmp.581850/9.train.py", line 156, in <module>
    main(args, config)
  File "/local/tmp.581850/9.train.py", line 138, in main
    trainer.fit(model, datamodule=data, ckpt_path=None)
  File "/envs/pytorch-env/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 696, in fit
    self._call_and_handle_interrupt(
  File "/envs/pytorch-env/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 650, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/envs/pytorch-env/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 735, in _fit_impl
    results = self._run(model, ckpt_path=self.ckpt_path)
  File "/envs/pytorch-env/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1102, in _run
    self.strategy.setup_environment()
  File "/envs/pytorch-env/lib/python3.9/site-packages/pytorch_lightning/strategies/ddp.py", line 157, in setup_environment
    self.setup_distributed()
  File "/envs/pytorch-env/lib/python3.9/site-packages/pytorch_lightning/strategies/ddp.py", line 209, in setup_distributed
    self._process_group_backend = self._get_process_group_backend()
  File "/envs/pytorch-env/lib/python3.9/site-packages/pytorch_lightning/strategies/ddp.py", line 216, in _get_process_group_backend
    or get_default_process_group_backend_for_device(self.root_device)
  File "/envs/pytorch-env/lib/python3.9/site-packages/pytorch_lightning/strategies/ddp.py", line 123, in root_device
    return self.parallel_devices[self.local_rank]
IndexError: list index out of range
```

**Reply by spirosbax** [reply]
I'm running into the exact same problem. I have followed the docs and created a SLURM config that matches trainer.num_nodes and trainer.devices with slurm nodes and tasks_per_node.

**Comment by Viditagarwal7479** [comment]
I am using torchrun and getting the same error when I am using
```sh
srun torchrun \
        --nnodes $SLURM_NNODES \
        --nproc_per_node $SLURM_NTASKS_PER_NODE \
        --rdzv_id $RANDOM \
        --rdzv_backend c10d \
        --rdzv_endpoint $head_node_ip:29500 \
        train_multi_ddp.py
```
Whereas this ones runs perfectly fine:
```sh
srun torchrun \
        --nnodes 2 \
        --nproc_per_node 1 \
        --rdzv_id $RANDOM \
        --rdzv_backend c10d \
        --rdzv_endpoint $head_node_ip:29500 \
        train_multi_ddp.py
```
I also verified the variables: 
```
echo $SLURM_NNODES
> 2
echo $SLURM_NTASKS_PER_NODE
> 1
```

The SLURM script configuration is:
```sh
#SBATCH -N 2
#SBATCH --tasks-per-node=1
#SBATCH --cpus-per-task=5
#SBATCH --gres=gpu:1
#SBATCH --job-name=multi_ddp
#SBATCH --output=train_multi_ddp.out
#SBATCH --error=train_multi_ddp.err
#SBATCH --partition=gpu
```
